File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cs
public class Solution {
    public string ReverseVowels(string s) {
        //  C++에서 내가 푼 방법과 다른 풀이인 솔루션 참고해서 작성한 코드
        string vowels = "aeiouAEIOU";
        int start = 0;
        int end = s.Length - 1;
        StringBuilder sb = new StringBuilder(s);
        while(start < end){
            while(start <end && vowels.IndexOf(sb[start]) == -1){
                start++;
            }
            while(start <end && vowels.IndexOf(sb[end]) == -1){
                end--;
            }
            char tmp = sb[start];
            sb[start] = sb[end];
            sb[end] = tmp;
            start++; 
            end--;
        }
        return sb.ToString();
        
        //  ============================
        //  https://leetcode.com/problems/reverse-vowels-of-a-string/solutions/3669092/easy-solution-with-stack
//         var str = new StringBuilder();
//         var l = new Stack<char>();

//         foreach(var i in s){
//             //  모음이 있으면 스택에 push
//             if("aeiouAEIOU".Contains(i)){
//                 l.Push(i);
//             }
//         }

//         foreach(var i in s){
//             //  모음이 있으면 스택에서 pop
//             //  LIFO 구조이기 때문에 s의 뒤쪽 모음부터 
//             //  str에 append한다
//             if("aeiouAEIOU".Contains(i)){
//                 str.Apeend(l.Pop());
//             }else{
//                 str.Append(i);
//             }
//         }

//         return str.ToString();
    }
}

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        vector<char> vec;
        for(int i = 0; i < s.length(); i++){
            if(s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u'
                || s[i] == 'A' || s[i] == 'E' || s[i] == 'I' || s[i] == 'O' || s[i] == 'U'){
                vec.push_back(s[i]);
            }
        }
        int index = vec.size() - 1;
        string ans = "";
        for(int i = 0; i < s.length(); i++){
            if(s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u'
                || s[i] == 'A' || s[i] == 'E' || s[i] == 'I' || s[i] == 'O' || s[i] == 'U'){
                ans += vec[index--];
            }else{
                ans += s[i];
            }
        }
        return ans;

        //  ===================================
        //  https://leetcode.com/problems/reverse-vowels-of-a-string/solutions/3723102/easy-solution-with-simple-explanation-using-two-pointer
        // string word = s;
        // int start = 0;
        // int end = s.length() - 1;
        // string vowels = "aeiouAEIOU";

        // while(start < end){
        //     while(start < end && vowels.find(word[start]) == string::npos){
        //         start++;
        //     }
        //     while(start < end && vowels.find(word[end]) == string::npos){
        //         end--;
        //     }
        //     swap(word[start], word[end]);
        //     start++;
        //     end--;
        // }
        // return word;
    }
};